File: python-main-Underwater Acoustic Channels Group-lizhengnan/examples/example_unpack.py
import numpy as np
import h5py
import matplotlib.pyplot as plt
from uwa_channels import unpack


if __name__ == "__main__":
    ## Load channel impulse responses
    channel = h5py.File("blue_1.mat", "r")

    ## Parammeters
    fs_delay = channel["params"]["fs_delay"][0, 0]
    fs_time = 20
    array_index = [0, 2, 4]

    ## Unpack the channel
    unpacked = unpack(fs_time, array_index, channel)

    ## Visualize
    delay_axis = np.arange(unpacked.shape[0]) / fs_delay
    time_axis = np.arange(unpacked.shape[2]) / fs_time
    plt.pcolor(
        delay_axis * 1e3,
        time_axis,
        20 * np.log10(np.abs(np.squeeze(unpacked[:, 0, :]))).T,
        vmin=-30,
        vmax=0,
    )
    plt.gca().invert_yaxis()  # Align with MATLAB
    plt.xlabel("Delay [ms]")
    plt.ylabel("Time [s]")

    plt.show()

# [EOF]

File: matlab-main/examples/example_unpack.m
% Demonstration of the unpack function.
%
% Author: Zhengnan Li
% Email : uwa-channels@ofdm.link
%
% License: MIT
%
% Revision history:
%   - Apr. 1, 2025: initial release.
%
%

clc;
clear;
close all;

%% Add the toolbox to the path
addpath('../src');

%% Load channel impulse responses
channel = load('blue_1.mat');
% channel = rmfield(channel, 'theta_hat');
% channel.f_resamp = 1.0002;

%% Parameters
fs_delay = channel.params.fs_delay;
fs_time = fs_delay * 0.005;
array_index = [1, 3, 5];

%% Unpack the channel
% Buffer fraction for impulse response padding. Increase these values if the
% impulse responses slide out of the window.
buffer_left = 0.1;
buffer_right = 0.1;
unpacked = unpack(fs_time, array_index, channel, buffer_left, buffer_right);

%% Visualize
delay_axis = (0:size(unpacked, 1) - 1) ./ fs_delay;
time_axis = (0:size(unpacked, 3) - 1) ./ fs_time;

figure
imagesc(delay_axis*1e3, time_axis, 20*log10(squeeze(abs(unpacked(:, 1, :))).'), [-30, 0])
xlabel('Delay [ms]')
ylabel('Time [s]')

%% Remove the toolbox from the path
rmpath('../src')

% [EOF]
